go/ephemeral/ephemeral.go

```diff
@@ -188,13 +193,236 @@ func GetOwnDeviceEKs(ctx context.Context, g *libkb.GlobalContext) ([]keybase1.De
 		// and skip them.
 		ageSecs := currentMerkleRoot.Ctime() - signedMerkleRoot.Ctime
 		if ageSecs > KeyLifetimeSecs {
-			g.Log.Debug("skipping expired deviceEK %s for device KID %s", signedMetadata.Kid, matchingDevice.KID)
+			g.Log.Debug("skipping stale deviceEK %s for device KID %s", verifiedMetadata.Kid, matchingDevice.KID)
 			continue
 		}
 
 		// This key is valid and current. Add it to the list we're about to return.
-		currentDeviceEKs = append(currentDeviceEKs, signedMetadata)
+		currentDeviceEKs[matchingDevice.DeviceID] = verifiedMetadata
 	}
 
 	return currentDeviceEKs, nil
 }
+
+// ===============
+// === UserEKs ===
+// ===============
+
+type UserEKSeed [EKSeedSize]byte
+
+func newUserEphemeralSeed() (seed UserEKSeed, err error) {
+	randomSeed, err := makeNewRandomSeed()
+	if err != nil {
+		return seed, err
+	}
+	return UserEKSeed(randomSeed), nil
+}
+
+func (s *UserEKSeed) DeriveDHKey() (key *libkb.NaclDHKeyPair, err error) {
+	derived, err := libkb.DeriveFromSecret(*s, libkb.DeriveReasonUserEKEncryption)
+	if err != nil {
+		return nil, err
+	}
+	keypair, err := libkb.MakeNaclDHKeyPairFromSecret(derived)
+	return &keypair, err
+}
+
+type UserEKBoxMetadata struct {
+	RecipientDeviceID   keybase1.DeviceID `json:"recipient_device_id"`
+	RecipientGeneration int               `json:"recipient_generation"`
+	Box                 string            `json:"box"`
+}
+
+func postNewUserEK(ctx context.Context, g *libkb.GlobalContext, sig string, boxes []UserEKBoxMetadata) error {
+	boxesJSON, err := json.Marshal(boxes)
+	if err != nil {
+		return err
+	}
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/user_ek",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args: libkb.HTTPArgs{
+			"sig":   libkb.S{Val: sig},
+			"boxes": libkb.S{Val: string(boxesJSON)},
+		},
+	}
+	_, err = g.GetAPI().Post(apiArg)
+	return err
+}
+
+func PublishNewUserEK(ctx context.Context, g *libkb.GlobalContext) (data keybase1.UserEkMetadata, err error) {
+	currentMerkleRoot, err := g.GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	if err != nil {
+		return data, err
+	}
+
+	// TODO: Read the actual generation from the userEK store.
```

Comment:
we'll also have to have retry logic similar to the deviceEK story if our local store doesn't have the latest generation for whatever reason